### DISCUSSION TITLE
Android Wear Support

### DISCUSSION THREAD

**Discussion by jones139** [OpenSeizureDetector/Android_Pebble_SD#69] [04_Devices]
User @AroonPro is working on providing a version of OpenSeizureDetector that can use an Android Wear device instead of a Garmin watch.  His version of the watch app is here:  https://github.com/AroonPro/AndroidWear_SD/tree/SD_AW/BR
We will probably have a separate companion app to deal with communication with the Android Wear watch, which then sends data to the main OpenSeizureDetector app - the same way that Garmin Connect works for Garmin devices.

**Comment by jones139** [comment]
There has been some interest in this lately, so I thought I should set out where I think we are with AndroidWear support, and what our options are.

I think we have two draft Implementations using different approaches:

**Analysis on Watch**
User @AroonPro has a version which he is using which (I believe) is doing the seizure detection algorithm on the watch (https://github.com/AroonPro/AndroidWear_SD) and sending the processed data back to the phone for alarm generation (which is the same as we do for the old Pebble watches).    I (Graham, @jones139) have not looked at what we need to do to get this working in V5.0.x)

** Analysis on Phone **
User @DylanGriffith has a prototype watch app which sends raw data back to the phone for processing (the same as we do for PineTime and Garmin watches).    Version 5.0.x of the [Android_Pebble_SD phone app](https://github.com/OpenSeizureDetector/Android_Pebble_SD/tree/beta) has support enabled for Android Wear messages so we should be able to get this working.


Should we go with one or both of these approaches?    Things to consider:

1. How is the battery life on the AndroidWear watches if we send data continuously?    If it is causing battery issues, then analysis on phone could be sensible - that is why we did it that way for Pebble watches.
2. Do some AndroidWear devices have their own SIM cards?   If that is the case we could avoid needing the phone app at all, and the AndroidWear implementation could be completely separate from the Android_Pebble_SD phone app.
3. Do we want to use Machine Learning algorithms on AndroidWear?   If so we should go with Analysis on Phone so we only have to implement the algorithm in one place, otherwise we will have to maintain several implementations for different devices.  Also, we are currently looking at using a Neural Network algorithm - can an AndroidWear device handle pyTorch models?

[Note that I know quite a few people are interested in AndroidWear support, so I  (Graham) will help include it in the main phone app, but I am not really interested in putting any effort into it myself - I think the main development drive should be in getting a machine learning algorithm fully implemented to reduce the false alarm rate from the system, rather than trying to support more devices, when we know that the low cost PineTime watch works]

**Reply by DylanGriffith** [reply]
Thanks @jones139 I've not had much time to look into this in a while. But I'll try to share my experience:

1. Battery was good enough the send interval was increased. Originally I was sending at 25Hz which drained the battery in a few hours. But sampling up to 10s or 30s should be fine for a full day.
2. Yes some do but I don't have one. The implementation I have is fully dependent on the mobile app and I wasn't really interested in building the complete app on Wear. I also don't know the battery usage for doing this kind of computation on the watch.
3. Yes we should use the best/tested algorithm. And yeah that's fine if it's on mobile.

I hope that I can get back to testing the AndroidWear app again soon but I stopped testing it because it was constantly sending false positives. And I didn't really get to the bottom of that. I was observing weird cases where an SMS would be sent but I never even heard it beep (and the beep settings were on as I heard them sometimes). All that was making it a problem to test and I didn't really know the codebase well enough to debug why. Possibly it was just my settings.

With that being said I think the biggest thing I'd want to know to put the effort into the AndroidWear app is how we can be sure the Machine Learning algorithm inputs are equivalent for Android Wear and Garmin. If all the training data is Garmin and there are some quirks that are different I'd be worried we risk either have false positives or no detection at all. False positives we can learn about but no detection at all seems to be a worse problem and I'd assume we need more data to test that. But maybe you have ideas about how to address this problem?

**Reply by jones139** [reply]
Hello Dylan,
I think implementing the algorithm on the watch is the simplest, we just need to make sure the data is the same as we get from the Garmin and PineTime watches.  That is:

1. Collect 125 accelerometer samples at 25 Hz (= 5 seconds of data).   On garmin I collect all 125 samples on the watch and send them in one go. On PineTime I collect 1 seconds worth of data and send that, and collect the 125 samples on the phone side instead.
2. The accelerometer values should be milli-g (so the vector magnitude is around 1000).    We stack the 3d accelerometer data into a single rawData3D array as [ x0, y0, z0, x1, y1, z1......x124, y124, z124] - see https://github.com/OpenSeizureDetector/Android_Pebble_SD/blob/beta/app/src/main/java/uk/org/openseizuredetector/datasource/SdDataSourceBLE2.java#L533, and then populate the SdData class as shown here:  https://github.com/OpenSeizureDetector/Android_Pebble_SD/blob/beta/app/src/main/java/uk/org/openseizuredetector/datasource/SdDataSourceBLE2.java#L558... and call doAnalysis() to actually process the data.

If the data is at the same 25Hz sample frequency and is in units of milli-g, then the algorithms are device agnostic - I see similar results using PineTime and Garmin watches, so we should be able to get AndroidWear working in the same way.   (when I first looked at Android Wear (it must be nearly 10 years ago!), the concern I had was that it was not sending data at a steady rate, as though the watch blocked it while it was doing something else - but I don't think @AroonPro has seen the problem, so the newer watches must have solved that issue.

To test the processing toolchain we could use our database of seizure data to make some 'standard' test data for you that you could have the watch send, rather than sending real data so you could convince yourself that it works - or just send obviously test data (Acceleration values of 1000, 10001, 10002 etc. that you can trace through the analysis to make sure it arrives correctly.     I am sure there is a 'proper' android testing way of doing this, but I'm not clever enough for that!

I'll reply to the comment by @AroonPro next - I think one thing we'll need to do is agree a 'Message Type' element in the AndroidWear messages to tell the phone app if it is a raw data packet (your version), or processed data (from the @AroonPro watch processing).

Thanks!
Graham.

**Reply by AroonPro** [reply]
Yes new galaxy watch have eSim. 
Also they have through bluetooth or wifi or router option in phone available. 

That now messages are running through the Google cloud.

That is why I really want the ssh connection to be active. 
We are too dependant on big tech. 

With ssh, it is low profile and secure.

**Reply by jones139** [reply]
@AroonPro , Are you saying that the communication between the watch and the phone is going via Google Cloud Services?    Surely not - that would imply you need an internet connection to have the watch and phone communicate?    If that is the case, I'd suggest dropping Android Wear altogether!

**Reply by AroonPro** [reply]
It is not exactly running through the cloud. That is how I see it works. 
With the ssh binding, the direct through approach Will  be better. 
Through Google messages on wear, it seems to be passed as Google cloud, but handled locally, before the cloud. 

For that, I need to examine the docs. 
For now, I try to rebuild, with SdData, SdServer, SdDataSource and osdUtil in uk.org.openseizuredetector. openseizuredetector class. Then there is only one point of truth.
However, with @jones139 pase of new versions, I need to push the project into the request. But health and trust issues hold me down.

**Reply by jones139** [reply]
The doucmentation here: https://android-wear-docs.readthedocs.io/en/latest/sync.html#:~:text=An%20application%20that%20runs%20on,First%20Wearable%20Message
Implies that it is a normal bluetooth communication without any cloud based layer - I suppose the way to check would be to switch off wifi and mobile data and check the watch and phone can still communicate.

**Reply by AroonPro** [reply]
Like I said.

It also uses the wifi or cloud routing.

This is a clever design

**Comment by AroonPro** [comment]
Certainly,

Our plan is to use the android test harness with real physical data sets as
benchmark.

Also, have I been working on stripping SdServer SdData and SdDataSource
from the android wear  and pebble_sd app and moving that into
openseizuredetector openseizuredetector
Then, no discussion afterwards upon changes upstream in pebble_sd /SdData.

Here comes a plus: wear watch becomes osd server.

With that, we can use a ssh connection to cloudflare uid
openseizuredetector openseizuredetector.co.uk for logging and with the same
address as network source display real_time alarms on pebble_sd app.
(this is the nuclear future feature, because it requires a certain
bandwidth. I.e. Fiberglass/4g+)

Met vriendelijke groet,

A. A. F. M. (Bram) Regtien
e: ***@***.***
m: +31614464152
a: Twentehof 13
    5709 KM Helmond

Op di 31 mrt 2026, 19:59 schreef Dylan ***@***.***>:

**Reply by jones139** [reply]
Hi Bram,
I don't think you need to change very much in the OSD phone app to achieve what you want.      If you process the data on the watch and send it to the phone with a flag to make it clear that it is pre-processed, we can handle it like we do data from the old Pebble watches (which also did on-watch processing).

Then the SdDataSourceAW on the phone app just needs to read the message, and determine if it is raw or processed data.   If it is raw data, we collate into a 5 second SdData instance and call doAnalysis() to process it.   If it is processsed data, we will bypass the analysis part and go straight to generating alarms.

I think the remote data access is a separate issue to AndroidWear isn't it?   It should not matter which data source we are using.    Maybe we should start another discussion about options for that, as we know a couple of people are using it over VPNs, and I think your idea will avoid the need for a VPN?

**Comment by AroonPro** [comment]
> Hi Bram,
> I don't think you need to change very much in the OSD phone app to achieve
> what you want. If you process the data on the watch and send it to the
> phone with a flag to make it clear that it is pre-processed, we can handle
> it like we do data from the old Pebble watches (which also did on-watch
> processing).
>
> Then the SdDataSourceAW on the phone app just needs to read the message,
> and determine if it is raw or processed data. If it is raw data, we collate
> into a 5 second SdData instance and call doAnalysis() to process it. If it
> is processsed data, we will bypass the analysis part and go straight to
> generating alarms.
>
> I think the remote data access is a separate issue to AndroidWear isn't
> it? It should not matter which data source we are using. Maybe we should
> start another discussion about options for that, as we know a couple of
> people are using it over VPNs, and I think your idea will avoid the need
> for a VPN?
>

This is about it. Where VPN is good, it is always a struggle with naming
servers.

Re binding through ssh, cloudless to new interface ipv6 at
openseizuredetector.co.uk, we have our automatic twin direction, zero
config data connection. With absolute privacy and ddos protection.
User compartination guaranteed.
In this configuration, group access to user database per account with
anonymised health data sharing enabled will be possible.

**Comment by jones139** [comment]
I get quite a lot of people asking about WearOS support, so I'll have a go with it - I have forked the @DylanGriffith watch app code so I can fiddle with it, and will have a go at integrating it with V5.1 of the Phone App this evening....   My fork is here:  https://github.com/OpenSeizureDetector/AndroidWear-watch-app

Maybe I should re-name it WearOS, because they seem to have dropped the AndroidWear name somewhere?

**Reply by DylanGriffith** [reply]
@jones139 sorry I forgot to update here. I actually wrote some more code on this a few weeks ago and pushed it to this separate branch https://github.com/DylanGriffith/openseizuredetector-watch-app/commits/rebuild-v2 .

I made improvements to the app where I could dismiss alerts from the app, which I considered to be crucial. But even with that after using it for 2 days straight with the machine learning algorithm (only) enabled I ended up with too many false positives for it to be usable. Basically it would go off whenever I was walking and often when I was typing. I couldn't figure out why and my biggest concern is that I'm not confident I have any way to collect realistic looking data. I'm sure I can tune the code so that it stops triggering false positives but without knowing how the sensors respond to a real seizure it seems I could never be certain that it actually works.

That's the main reason I kinda gave up again on this project. That and battery life wasn't great. With the batching it was around 8 hours of battery compared to around 24 hours without the app.

**Reply by jones139** [reply]
Thanks for confirming @DylanGriffith - I was about to have a look to see if you had different branches because I thought the version I just built might be missing something.

So on the positive side, it sounds like you have the transfer from the watch to the phone working, which is good (query about accuracy of data - see below), but 8 hours battery life is no good at all - this is actually why I abandoned my first go at AndroidWear - I could get over 15 hours off Garmin watches (24 on some), and we now get well over 24 hours from PineTime.    So battery life would be the main thing to look at.

On the false alerts, there are couple of possibilities:

1. The issue is with the ML algorithm itself - another user has reported problems with typing setting it off - this could be because most of the typing data in the database comes from me, and if we tend to type at different rates, that could be an issue which is fairly easily fixed by recording more typing data and re-training the ML model.    Walking is more of a surprise - it goes off for me if I walk with my arms still (because I am carrying something) but not if I walk normally with a bit of arm swing.   Again this could just be a matter of training with more false alarm data.
2. The other possibility is that the sample frequency coming from the WearOS device is not 25 Hz - on the Phone data source I can only get 50Hz so downsample to 25 Hz to be compatible with Garmin and PineTime watches.    The way you could tell this is if you put the phone app into 'advanced mode' (or rather not basic mode), then the UI shows the time since the last update from the watch, and the time for the last 125 samples to be collected.    The time for 125 samples should be 5 seconds +/- the accuracy of the accelerometer sample frequency setting, and time should count up to 4 and reset back to zero.    I don't have a wearOS watch so can't test it - if you get chance will you have a look and see what these timers are doing please?

Another check you can do is if you enable the original 'OSD' deterministic algorithm, the main UI shows a power spectrum - if you imagine being very cold and shivering, you should see a peak at about 4-5 Hz - if it is lower than this it is a sign that it is sampling quicker.

I was about to buy myself a Galaxy Watch 7 to use for testing, but given your battery life report, I might not bother!

I think that @AroonPro had reported better battery life than you had though?

Regards

Graham.

**Reply by DylanGriffith** [reply]
> it goes off for me if I walk with my arms still (because I am carrying something) 

@jones139 it's very possible this was the problem I was having. I was walking while holding my phone a lot and it was going off a lot.

A lot of this does make me think we'd really need a lot more data to make the ML algorithm reliable though. It also feels kinda surprising because I'd assume the signature of seizure data should look so different to walking that there really should be a different data analysis technique needed here but I don't know how to validate this hypothesis without a wider research project.

**Reply by jones139** [reply]
@DylanGriffith, I think I should start a separate discussion about detection algorithms, but to answer your point:

We always need more data :), but what I was trying to do with this current ML algorithm (run_24) is to get something which is as good as the deterministic OSD algorithm for seizure detection reliability, and gives some improvement in false alarms.    This one achieves that target, so I decided to publish it and update the android app to use it, so we (or I!), can get used to trusting an ML model, and then iterate its training to improve it as we get more data.

Now, the model I have used is one inspired by [this paper] (https://github.com/OpenSeizureDetector/OpenSeizureDatabase/blob/main/Publications/Epilepsia%20-%202025%20-%20Spahr%20-%20Deep%20learning%20based%20detection%20of%20generalized%20convulsive%20seizures%20using%20a%20wrist%E2%80%90worn.pdf).   They got successful results of a relatively small amount of seizure data, using a surprisingly complex neural network - I had previously shied away from such a complex network because it is really, really easy to over-fit it to your data so that it gives good results that are not generalised to other data.     This is addressed by making it quite 'forgetful' during training so we can have a complex model that is not over-fitted.

The key thing is that it was trained against our known seizure data and false alarms and normal-daily-activity data from the database, but the false alarms were mostly generated by the original OSD algorithm.    What we are seeing now is that this ML model actually false alarms for other things, which the original model did not - so what we need to do is collect false alarm data from this model, and re-train it to teach it that these events are not false alarms.

I have just set my computer loose on the current database, which includes false alarms.   I should really do a bit more sense checking of the data, because there might be some seizures reported for the wrong time so might not be real seizure data, but this is a good test of how good it is without any manual intervention.   It should finish overnight so tomorrow I will compare it to the raw statistical results from run24 and see how they compare.   If this one is rubbish, then I'll need to do the manual filtering of the data.

Will let you now how it looks tomorrow sometime.......

Graham.

**Reply by jones139** [reply]
My computer is nice and fast, so got the result tonight.....   Just throwing all of the data at it gave results that look fairly similar to the existing ML model - it does better than the deterministic OSD algorithm for seizure detection and a better false alarm rate - but still quite a lot of false alarms.
I think one issue is that we have a new contributor that has a lot of seizures, but I am suspicious about the classification of some of them, so they may be degrading false alarm performance.
I'll do a sanity check and tidy the database and try again tomorrow....

**Reply by DylanGriffith** [reply]
> We always need more data

@jones139 thanks. I just wanted to show appreciation for the effort you're putting into this and the context about your approach to improving the algorithm is very helpful.

Thanks again for your effort and I'll probably come back around to trying out the next version in a couple of months. I may need to try to setup the data sharing stuff this time so you can have some of my false positive data from walking around to see if it helps.

**Comment by AroonPro** [comment]
At my own repo https://github.com/AroonPro/AndroidWear_SD I have a new core
routing embedded. It has all the new (to last commit date modules from the
main app.

Pretty straight forward, battery time quite acceptable.

However, too trigger happy and android studio gemini forgets too much parts
toe place back
Best Regards,

A. A. F. M. (Bram) Regtien
e: ***@***.***
m: +31614464152
a: Twentehof 13
    5709 KM Helmond
 Student Rechtswetenschappen (LL.M. traject)
 Student Systeem Architect (EVC-traject MSc/ir. niveau)

Open Universiteit
 e: ***@***.***
Uit de familie Re
<https://www.genealogieonline.nl/stamboom_hendriks_regtien/I6252.php>(ch
<https://www.genealogieonline.nl/stamboom_hendriks_regtien/I1836.php>/g
<https://www.genealogieonline.nl/stamboom_hendriks_regtien/I1945.php>)tien /
Sargentini
<https://www.genealogieonline.nl/stamboom_hendriks_regtien/I8090.php> / di
Regoli <https://www.genealogieonline.nl/stamboom_hendriks_regtien/I799.php>-
van Hoof

Op za 8 aug 2026, 14:47 schreef Dylan ***@***.***>:

**Comment by jones139** [comment]
@DylanGriffith, @AroonPro - which watches are you both using?   I am wondering if it is a hardware or a software difference - if the version of @AroonPro can get of the order 24 hours battery life on a similar watch, it could be as simple as how the accelerometer is being used - I know that when I have written code to access the accelerometer directly, you can either request every measurement, or have it buffer it and transfer several measurements in one go to save CPU cycles.

**Reply by DylanGriffith** [reply]
> which watches are you both using?

I'm using the Pixel Watch 4

**Comment by jones139** [comment]
Well, I am confused now :).   I asked Github Copilot (Claude Haiku model) to compare the sensor handling between the @DylanGriffith and the @AroonPro implementations.....and it is concluding that the @AroonPro one should be more battery intensive (because of HR sensor handling).   I haven't checked myself, but I find the Claude models quite reliable for this sort of analysis...
Output below for reference:


I'll analyze the accelerometer and sensor setup in the second repository to compare power consumption characteristics.Let me explore the repository structure first:# Sensor Handling Comparison: AroonPro/AndroidWear_SD vs OpenSeizureDetector/AndroidWear-watch-app

## Executive Summary

The second app (OpenSeizureDetector/AndroidWear-watch-app, rebuild-v2 branch) has **significantly better battery efficiency** compared to the first app (AroonPro/AndroidWear_SD). The primary differences relate to **heart rate sensor handling** and messaging architecture rather than accelerometer setup. The second app avoids substantial battery drain through simpler, more conservative HR sensor management.

---

## Accelerometer Configuration Comparison

### Similarities (No Significant Difference)

Both apps use virtually identical accelerometer setups:

| Parameter | AroonPro/AndroidWear_SD | OpenSeizureDetector rebuild-v2 |
|-----------|-------------------------|--------------------------------|
| **Sample Rate** | 25 Hz | 25 Hz |
| **Sample Period** | 40,000 μs | 40,000 μs (1,000,000 / 25) |
| **Analysis Window** | 10 seconds (250 samples) | 1 second batches (25 samples) |
| **Buffering Strategy** | On-chip hardware buffering with 3× latency multiplier | No explicit latency parameter—relies on default buffering |
| **Data Processing** | FFT-based frequency analysis (3–8 Hz alarm ROI) | Minimal—batches and sends raw samples to phone |
| **Downsampling** | Applied during collection | N/A (sends raw data) |

**Conclusion**: The accelerometer handling is **essentially equivalent** in battery impact. Both use 25 Hz sampling with hardware batching to reduce wake-ups.

---

## Heart Rate Sensor: Critical Difference

### AroonPro/AndroidWear_SD (High Battery Drain)

```java
// bindSensorListeners() - Line 989-990
mHeartSensor = mSensorManager.getDefaultSensor(Sensor.TYPE_HEART_RATE);
mSensorManager.registerListener(this, mHeartSensor, 
    (int) Constants.GLOBAL_CONSTANTS.getMaxHeartRefreshRate);
```

**Issue #1: Aggressive HR Sampling**
- Calculates `getMaxHeartRefreshRate = (1d / (maxHeartRefreshRate / 60d)) * 1000`
- With `maxHeartRefreshRate = 300 bpm` (unlikely but coded for), this yields **12 ms sampling interval** (83 Hz equivalent)
- Registers HR sensor with this microsecond-level parameter as the **sampling rate**, not buffering latency

**Issue #2: No Latency/Buffering Parameter**
- The `registerListener()` call uses only the sample rate parameter
- Omits the maximum report latency parameter, meaning the sensor delivers **every single reading to the app immediately**
- No on-chip batching; the CPU wakes up ~83 times per second for HR events alone

**Issue #3: Multiple Redundant HR Registrations**
```java
// Lines 1000, 1003, 1005
mHeartSensor = mSensorManager.getDefaultSensor(Sensor.TYPE_HEART_RATE);
mSensorManager.registerListener(this, mHeartSensor, 
    (int) Constants.GLOBAL_CONSTANTS.getMaxHeartRefreshRate);  // First registration

mHeartBeatSensor = mSensorManager.getDefaultSensor(Sensor.TYPE_HEART_BEAT);
mSensorManager.registerListener(this, mHeartSensor,  // BUG: registers mHeartSensor again!
    (int) Constants.GLOBAL_CONSTANTS.getMaxHeartRefreshRate, 
    (int) Constants.GLOBAL_CONSTANTS.getMaxHeartRefreshRate * 3);

mBloodPressure = mSensorManager.getDefaultSensor(Sensor.TYPE_PRESSURE);
mSensorManager.registerListener(this, mHeartSensor,  // BUG: registers mHeartSensor AGAIN!
    SensorManager.SENSOR_DELAY_UI);
```

**The app registers the heart rate sensor THREE times** with different sampling parameters. This causes:
- Multiple competing event streams to the same listener
- Unpredictable behavior (last registration wins on some devices)
- Massive redundant CPU wake-ups

**Issue #4: Continuous Data Processing**
```java
// onSensorChanged() - Lines 1167-1186
if (event.sensor.getType() == Sensor.TYPE_HEART_RATE && event.values.length > 0) {
    int newValue = Math.round(event.values[0]);
    if (mSdData.mHR != newValue && newValue != 0) {
        heartRates.add(mSdData.mHR);
        // ... adds to history every change
    }
    checkAlarm();  // Runs on EVERY HR event
    if(mHeartRatesCount % 10 == 0) {
        sendMessage(...);  // Sends data frequently
    }
}
```

Each HR sample triggers `checkAlarm()` and conditional messaging, compounding the wake-up cost.

---

### OpenSeizureDetector rebuild-v2 (Low Battery Drain)

```kotlin
// startSensors() - Lines 146-151
hrSensor?.let {
    sensorManager.registerListener(this, it, SensorManager.SENSOR_DELAY_NORMAL)
}
```

**Advantage #1: Conservative Sampling**
- Uses `SensorManager.SENSOR_DELAY_NORMAL` (typical delay ~200 ms, ~5 Hz)
- No latency parameter specified—defaults to immediate delivery, but at a much lower event rate
- **~17× fewer HR events** compared to AroonPro's 83 Hz

**Advantage #2: Single Registration**
- Registers the HR sensor **once** with a standard delay constant
- No conflicting registrations; no redundant CPU wake-ups

**Advantage #3: Batched Transmission**
```kotlin
// Lines 231-239, 412-425
private fun sendHrData() {
    if (latestHr <= 0) return
    try {
        val json = JSONObject().put("hr", latestHr)
        sendMessage(MessagePaths.HR_DATA, json.toString().toByteArray(StandardCharsets.UTF_8))
    }
}

// Periodic tick sends HR every 5 seconds
private val periodicTickRunnable = object : Runnable {
    override fun run() {
        sendHrData()
        handler.postDelayed(this, PERIODIC_TICK_MS)  // 5000 ms
    }
}
```

HR data is **batched and sent on a 5-second timer**, not on every sensor event. This:
- Reduces wireless messaging overhead
- Keeps the phone connection path predictable
- Minimizes CPU wake-ups

**Advantage #4: Minimal Per-Event Processing**
```kotlin
// onSensorChanged() - Lines 156-161
Sensor.TYPE_HEART_RATE -> {
    val hr = event.values[0].toInt()
    if (hr > 0) {
        latestHr = hr
        _uiState.update { it.copy(heartRate = hr) }
    }
}
```

Simply caches the latest value; no per-event alarm checking, no frequency averaging, no immediate messaging.

---

## Accelerometer Data Processing Comparison

| Aspect | AroonPro | OpenSeizureDetector |
|--------|----------|-------------------|
| **Processing** | On-watch FFT analysis (~250 samples) | Send raw batches to phone (25 samples/sec) |
| **Analysis Location** | Watch (CPU-intensive) | Phone (more efficient) |
| **Data Sent** | Aggregated spectrum + power metrics | Raw sensor values |
| **Frequency** | ~Every 10 seconds | ~Every 1 second (one batch) |
| **Energy Cost** | Higher (FFT computation on watch) | Lower (watch only buffers/sends) |

---

## Summary of Battery Impact

### **Heart Rate Sensor (Dominant Factor)**

| Issue | AroonPro | OpenSeizureDetector | Impact |
|-------|----------|-------------------|--------|
| **Sampling Rate** | ~83 Hz (12 ms interval) | ~5 Hz (200 ms typical) | **17× fewer HR interrupts** |
| **Event Batching** | Per-event alarm checks | 5-second batch timer | **Reduced messaging** |
| **Registrations** | 3× (buggy redundancy) | 1× (clean) | **No duplicate wake-ups** |
| **Processing** | Continuous (every event) | Cached value only | **Minimal per-event work** |

### **Accelerometer (Similar)**
- Both use 25 Hz with hardware buffering
- Minor difference: AroonPro processes on-watch; OpenSeizureDetector delegates to phone
- Impact: Negligible compared to HR differences

---

## Conclusion

**The OpenSeizureDetector rebuild-v2 app will drain battery significantly more slowly**, primarily due to:

1. **Conservative HR sampling** (5 Hz vs. 83 Hz) → ~17× fewer CPU wake-ups
2. **No redundant sensor registrations** → No conflicting event streams
3. **Batched HR transmission** (5-second intervals) → Fewer wireless operations
4. **Minimal per-event processing** → No per-sample alarm checks

The AroonPro app's battery drain is driven almost entirely by its aggressive, redundantly-registered heart rate sensor configuration, not by accelerometer handling. For a wearable device running continuous seizure detection, this difference could extend battery life from **12-18 hours to 24-36+ hours**.

**Comment by jones139** [comment]
I have ordered my self a Samsung Watch 7 to play with, so will get that on Monday.   In the meantime I have made a release build of @DylanGriffith version here:  https://github.com/OpenSeizureDetector/AndroidWear-watch-app/blob/rebuild-v2/releases/app-release-V0.0.1.apk in case anyone wants to try it.

But I don't know how you install an APK on a watch, so if anyone can suggest any instructions to add to it, please let me know!

**Comment by jones139** [comment]
So I have set up my Samsung Watch 7 to do continuous heart rate monitoring and other continuous options like O2 saturation and stress.   I set the display to switch off and only switch on when you rotate your wrist (because I don't want it glowing in the dark).   I have worn it for a day now and noted the following:

1. Battery consumption, so ok, but not great - It goes from 100% to 50% in about 20 hours, so around 40 hours battery life, without running the OSD app.
2. The optical heart rate monitor seems to struggle to accept that my heart rate is so low and sometimes jumps to an unrealistically high reading (e.g. 80 bpm when resting rather than 45-50 bpm)  It looks like someone else has an issue with the Samsung Optical HRM too - see https://eu.community.samsung.com/t5/wearables/major-heart-rate-accuracy-issue-on-galaxy-watch-ultra-2025-40/m-p/14548151#M77631
3. There are annoying 'well done, you've got up and moved' messages that keep popping up that I'll have to switch off somehow.
4. The EEG function seems surprsingly good, so I can check my dicky ticker more easily that with my Kardia ecg - but when my heart rate is less than 50 bpm it refuses to classify it, even though the trace looks fine - I'll have to work out how to tell it that 45-55 is about right for me when resting....
5. I seem to be able to install an app on it from Android Studio using Wifi Debugging, provided I stop the screen going to sleep, otherwise the wifi shuts down - the easiest way seems to be to put it on the charger.

So it seems like quite a nice device, and not too expensive (I paid £130), but the 40 hour battery life before we start with continuous accelerometer data streaming is a bit of a concern.

**Comment by jones139** [comment]
I now have the @DylanGriffith version running alongside V5.1.0b of the android app - will see how much battery it uses overnight.....

**Comment by jones139** [comment]
Well, very disappointing results with the Samsung Galaxy Watch 7........

1. Battery drained from 100% to 11% in about 6 hours, so battery life is well below the 12 hours that I consider the minimum to be useful.
2. The data transfer seemed uneven - rather than a consistent 5 seconds update, sometimes I saw as little as 3.5, other times as much as 8 seconds, with the watch and phone very close together.    This is either a the sampling frequency varying, or more likely re-tries on the data transfer for some reason.
3. The watch configuration was the same as for my test without OSD running the day before, which gave about 40 hours battery life, so it is the OSD app that is causing the battery drain.

Maybe the battery and data transfer issue might be connected - maybe it is a noisy transfer and it is doing a lot of retries to ensure data integrity?

But really the battery life is so far from being useful, and the unreliable heart rate monitor means I don't see much point in persevering to get this working better.     But if someone else would like to get it working, please try - I'll add the specific points I have identified as issues against https://github.com/OpenSeizureDetector/AndroidWear-watch-app so anyone wanting to try knows where we have got with it.

<img width="320"  alt="Screenshot_20260814_042657_OpenSeizureDetector" src="https://github.com/user-attachments/assets/b660a873-d0ae-47f0-8ee1-afc320c6e73e" />

**Comment by jones139** [comment]
https://github.com/OpenSeizureDetector/AndroidWear-watch-app/issues/3#issuecomment-5333444442

- It appears that the original version was not using hardware batching of accelerometer data, so I have tried implementing it (well, I told Google Gemini to implement it and it has done something sensible looking).    Will test it tonight to see if it does any better.....

(Commit https://github.com/OpenSeizureDetector/AndroidWear-watch-app/commit/2872f406a0647074b6af9f1bddeb0b282c5d2656)

**Reply by AroonPro** [reply]
Can you check my last build? 

I have used the libraries of osd main app.

These were shifted into root neighbor folder. 

It is efficient if I update as soon as I can. 

In the meantime I already have implemented ssh tunnel for use in esim setup.

**Reply by jones139** [reply]
I haven't checked your version yet, @AroonPro, but I'll have a look, because the change I made had a small benefit, but does not get us anywhere near 24 hours battery life.   I have had a look at this version and the sensors are running continuously - and the battery life is ok when the phone companion app is not running, so it is not the sensor handling that is causing the battery drain.   It implies it is the data transfer from the watch to the phone, which I find very surprising.      How are you doing the watch->phone communication?   Maybe your data transfer method is more efficient.

**Reply by jones139** [reply]
@AroonPro, Please confirm which repository and branch is your latest version - I can't quite decide if it is SD_AW/BR or another branch.
Thanks!

**Reply by AroonPro** [reply]
You have found the correct branch. 


Note: due to creation of uk.org.openseizuredetector.openseizuredetector, the wear app needs open seizure detector root in android projects. 

Then update root project with pebble sd modules. 

Then ask gemini to split function if running package contains .aw . Upon that, ui updates only portray update time and Hr, over x samples. Otherwise push accordingly original phone update scheme

**Comment by jones139** [comment]
**Current State of Development (Aug 2026)**

I had a bit of time while I was on holiday with only poor internet connection, so I did some work on the WearOS seizure detection implementation (because I couldn't log into my big computer to do ML algorithm training...).

- I have taken the light weight watch app developed by @DylanGriffith and made a repository under the OpenSeizureDetector project (https://github.com/OpenSeizureDetector/WearSD).  It is all in the 'main' branch to avoid any confusion.
- It is not published as an executable, so to run it, build it from source and install it using wifi debugging.
- The biggest issue with it is battery drain - we only get about 8 hours of battery life (https://github.com/OpenSeizureDetector/WearSD/issues/3).    This seems to be related to the communications between the watch and phone, because the battery life is more like 24 hours if we do not run the OpenSeizureDetector phone app, so the communications do not happen.
- Another issue I have is a general one with the Galaxy Watch 7, which is it seems to disconnect from the phone, even if the phone is close to the watch - I have not investigated this to see if the OSD communications still get through, but it is unlikely that they will, so this will affect the reliability of the system as a seizure detector.

**I'm not going to do any more on this for now** (I want to work on the ML algorithm to improve the PineTime and Garmin performance), but if someone wants to try to get it to a production state, some suggestions:

1. Do a simple test of disabling the phone communications to check that it is the communications that is causing the excessive battery drain.
2. Look at the @AroonPro version to see what that is doing differently that gives the better battery life that he reports - maybe it is a different way of communicating with the phone which is more energy efficient?
3. Look at the WearOS libraries to see if there are some battery usage data available, and log this to determine what is using battery.
4. Try it for a while to check if the watch-phone connection reliability is actually useable.

Thanks to everyone who has contributed to get it this far!

Graham.
